File: step_03_arrays/3.2_medium/P011_set_matrix_zeroes/solution.py
"""
Problem: Set Matrix Zeroes
Difficulty: MEDIUM | XP: 25
LeetCode: 73

Given an m x n integer matrix, if an element is 0, set its entire row and
column to 0. Do this in-place.
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(m*n*(m+n))  |  Space: O(m*n)
# ============================================================
def brute_force(matrix: List[List[int]]) -> None:
    """
    Copy the matrix. For each 0 in the original, mark the entire row and
    column in the copy. Then overwrite the original with the copy.
    Avoids the problem of marking new zeros and then incorrectly zeroing
    additional rows/columns based on those new zeros.
    """
    m, n = len(matrix), len(matrix[0])
    copy = [row[:] for row in matrix]

    for i in range(m):
        for j in range(n):
            if copy[i][j] == 0:
                # Zero out row i in the original
                for c in range(n):
                    matrix[i][c] = 0
                # Zero out column j in the original
                for r in range(m):
                    matrix[r][j] = 0


# ============================================================
# APPROACH 2: OPTIMAL — Row/Column Boolean Sets
# Time: O(m*n)  |  Space: O(m+n)
# ============================================================
def optimal(matrix: List[List[int]]) -> None:
    """
    First pass: record which rows and columns contain zeros in two sets.
    Second pass: zero out every cell whose row or column is in those sets.
    Two passes, O(m+n) extra space.
    """
    m, n = len(matrix), len(matrix[0])
    zero_rows = set()
    zero_cols = set()

    for i in range(m):
        for j in range(n):
            if matrix[i][j] == 0:
                zero_rows.add(i)
                zero_cols.add(j)

    for i in range(m):
        for j in range(n):
            if i in zero_rows or j in zero_cols:
                matrix[i][j] = 0


# ============================================================
# APPROACH 3: BEST — In-Place O(1) Space Using First Row/Column as Markers
# Time: O(m*n)  |  Space: O(1)
# ============================================================
def best(matrix: List[List[int]]) -> None:
    """
    Use the first row and first column of the matrix itself as marker arrays.
    - matrix[0][j] == 0 means column j should be zeroed.
    - matrix[i][0] == 0 means row i should be zeroed.
    Need an extra boolean to track if the first row itself had a zero,
    and another for the first column.
    """
    m, n = len(matrix), len(matrix[0])

    # Check if first row/column originally contain zeros
    first_row_has_zero = any(matrix[0][j] == 0 for j in range(n))
    first_col_has_zero = any(matrix[i][0] == 0 for i in range(m))

    # Use first row and column as markers for the rest of the matrix
    for i in range(1, m):
        for j in range(1, n):
            if matrix[i][j] == 0:
                matrix[i][0] = 0  # mark row
                matrix[0][j] = 0  # mark column

    # Zero out cells based on markers (skip first row/col for now)
    for i in range(1, m):
        for j in range(1, n):
            if matrix[i][0] == 0 or matrix[0][j] == 0:
                matrix[i][j] = 0

    # Handle first row
    if first_row_has_zero:
        for j in range(n):
            matrix[0][j] = 0

    # Handle first column
    if first_col_has_zero:
        for i in range(m):
            matrix[i][0] = 0


def print_matrix(matrix):
    for row in matrix:
        print(" ", row)


if __name__ == "__main__":
    print("=== Set Matrix Zeroes ===")

    import copy

    matrix1 = [[1, 1, 1], [1, 0, 1], [1, 1, 1]]
    m1a, m1b, m1c = copy.deepcopy(matrix1), copy.deepcopy(matrix1), copy.deepcopy(matrix1)
    print("Input:"); print_matrix(matrix1)
    brute_force(m1a); print("Brute:"); print_matrix(m1a)
    optimal(m1b);     print("Optimal:"); print_matrix(m1b)
    best(m1c);        print("Best:"); print_matrix(m1c)

    matrix2 = [[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]]
    m2a, m2b, m2c = copy.deepcopy(matrix2), copy.deepcopy(matrix2), copy.deepcopy(matrix2)
    print("\nInput:"); print_matrix(matrix2)
    brute_force(m2a); print("Brute:"); print_matrix(m2a)
    optimal(m2b);     print("Optimal:"); print_matrix(m2b)
    best(m2c);        print("Best:"); print_matrix(m2c)

File: step_03_arrays/3.2_medium/P011_set_matrix_zeroes/Solution.java
/**
 * Problem: Set Matrix Zeroes
 * Difficulty: MEDIUM | XP: 25
 * LeetCode: 73
 *
 * Given an m x n integer matrix, if an element is 0, set its entire row
 * and column to 0. Do this in-place.
 *
 * @author DSA_Nova
 */
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE
    // Time: O(m*n*(m+n))  |  Space: O(m*n)
    // ============================================================
    /**
     * Copy the matrix. Scan the copy for zeros. For each zero found,
     * zero out the entire row and column in the original matrix.
     * The copy prevents cascade: if we zeroed the original while scanning,
     * newly introduced zeros would incorrectly trigger further zeroing.
     */
    static class BruteForce {
        public void setZeroes(int[][] matrix) {
            int m = matrix.length, n = matrix[0].length;
            int[][] copy = new int[m][n];
            for (int i = 0; i < m; i++)
                copy[i] = Arrays.copyOf(matrix[i], n);

            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (copy[i][j] == 0) {
                        for (int c = 0; c < n; c++) matrix[i][c] = 0;
                        for (int r = 0; r < m; r++) matrix[r][j] = 0;
                    }
                }
            }
        }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL — Row/Column Boolean Sets
    // Time: O(m*n)  |  Space: O(m+n)
    // ============================================================
    /**
     * First pass: collect all row and column indices that contain zeros.
     * Second pass: zero out any cell whose row or column index is in those sets.
     */
    static class Optimal {
        public void setZeroes(int[][] matrix) {
            int m = matrix.length, n = matrix[0].length;
            Set<Integer> zeroRows = new HashSet<>();
            Set<Integer> zeroCols = new HashSet<>();

            for (int i = 0; i < m; i++)
                for (int j = 0; j < n; j++)
                    if (matrix[i][j] == 0) {
                        zeroRows.add(i);
                        zeroCols.add(j);
                    }

            for (int i = 0; i < m; i++)
                for (int j = 0; j < n; j++)
                    if (zeroRows.contains(i) || zeroCols.contains(j))
                        matrix[i][j] = 0;
        }
    }

    // ============================================================
    // APPROACH 3: BEST — In-Place O(1) Space (First Row/Col as Markers)
    // Time: O(m*n)  |  Space: O(1)
    // ============================================================
    /**
     * Use matrix[0][j] as the marker for column j, and matrix[i][0] as the
     * marker for row i. Two extra booleans track whether the first row and
     * first column themselves originally contained zeros (since we overwrite them).
     *
     * Steps:
     * 1. Check if row 0 or col 0 has any original zeros.
     * 2. Scan rows 1..m-1, cols 1..n-1: if matrix[i][j]==0, set matrix[i][0]=0 and matrix[0][j]=0.
     * 3. Use markers to zero out rows 1..m-1 and cols 1..n-1.
     * 4. Apply the original first-row/first-col zero flags.
     */
    static class Best {
        public void setZeroes(int[][] matrix) {
            int m = matrix.length, n = matrix[0].length;

            boolean firstRowZero = false, firstColZero = false;
            for (int j = 0; j < n; j++) if (matrix[0][j] == 0) firstRowZero = true;
            for (int i = 0; i < m; i++) if (matrix[i][0] == 0) firstColZero = true;

            // Use first row/col as markers
            for (int i = 1; i < m; i++)
                for (int j = 1; j < n; j++)
                    if (matrix[i][j] == 0) {
                        matrix[i][0] = 0;
                        matrix[0][j] = 0;
                    }

            // Zero out based on markers
            for (int i = 1; i < m; i++)
                for (int j = 1; j < n; j++)
                    if (matrix[i][0] == 0 || matrix[0][j] == 0)
                        matrix[i][j] = 0;

            // Handle first row
            if (firstRowZero)
                for (int j = 0; j < n; j++) matrix[0][j] = 0;

            // Handle first column
            if (firstColZero)
                for (int i = 0; i < m; i++) matrix[i][0] = 0;
        }
    }

    static void printMatrix(int[][] m) {
        for (int[] row : m) System.out.println("  " + Arrays.toString(row));
    }

    static int[][] deepCopy(int[][] m) {
        int[][] copy = new int[m.length][];
        for (int i = 0; i < m.length; i++) copy[i] = Arrays.copyOf(m[i], m[i].length);
        return copy;
    }

    public static void main(String[] args) {
        System.out.println("=== Set Matrix Zeroes ===");

        int[][] matrix1 = {{1,1,1},{1,0,1},{1,1,1}};
        System.out.println("Input:"); printMatrix(matrix1);
        int[][] m1a = deepCopy(matrix1), m1b = deepCopy(matrix1), m1c = deepCopy(matrix1);
        new BruteForce().setZeroes(m1a); System.out.println("Brute:");   printMatrix(m1a);
        new Optimal().setZeroes(m1b);    System.out.println("Optimal:"); printMatrix(m1b);
        new Best().setZeroes(m1c);       System.out.println("Best:");    printMatrix(m1c);

        int[][] matrix2 = {{0,1,2,0},{3,4,5,2},{1,3,1,5}};
        System.out.println("\nInput:"); printMatrix(matrix2);
        int[][] m2a = deepCopy(matrix2), m2b = deepCopy(matrix2), m2c = deepCopy(matrix2);
        new BruteForce().setZeroes(m2a); System.out.println("Brute:");   printMatrix(m2a);
        new Optimal().setZeroes(m2b);    System.out.println("Optimal:"); printMatrix(m2b);
        new Best().setZeroes(m2c);       System.out.println("Best:");    printMatrix(m2c);
    }
}
